warp10/src/main/java/io/warp10/script/functions/ASREGS.java

```diff
@@ -242,16 +242,40 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
             }
           } else if (symbol instanceof ENDLIST) {
             int idx = i - 2;
+            int nbOfReg = 0; // Keeps track of the number of registers in this list
             while (idx >= 0 && !(statements.get(idx) instanceof MARK)) {
               Object stmt = statements.get(idx);
               if (stmt instanceof String) {
                 Integer regno = varregs.get(stmt);
                 if (null != regno) {
                   statements.set(idx, (long) regno);
+                  nbOfReg++;
                 }
+              } else if (stmt instanceof Long) {
+                nbOfReg++;
               }
               idx--;
-            }            
+            }
+
+            // Further optimization: if the list contains only registers, replace by POPRs which is much faster.
+            // For instance, replace [ 3 7 9 ] STORE by NOOP POPR9 POPR7 POPR3 NOOP NOOP.
+            int listLength = i - idx - 2;
+            if (nbOfReg == listLength) {
+              statements.set(idx, NOOP); // replace MARK
+              statements.set(i - 1, NOOP); // replace ENDLIST
+              statements.set(i, NOOP); // replace STORE
+
+              // As we must flip the order of the list, we must store the registers first.
+              int[] regInList = new int[listLength];
+              for (int listIndex = 0; listIndex < listLength; listIndex++) {
+                regInList[listIndex] = ((Long) statements.get(idx + 1 + listIndex)).intValue();
+              }
+
+              // Replace register number by POPRs. Be careful, we flip the list order!
```

Comment:
What happens if you initially had 1 2 3 [ 1 2 2 ] STORE? It seems that register 2 will end up with value 2 instead of 3.

Reply:
Good question, I didn't think about this case.
Turns out
```
  1 2 3 [ 1 2 2 ] STORE
  2 LOAD
```
Leave `2` on the stack as `STORE` iterates through the array backwards poping and storing. So the behavior is the same with or without `ASREGS`.

However there is an other bug, I'm investigating.

Reply:
I think we should consider fixing STORE in the process, 1 2 3 [ 1 2 2 ] STORE should really leave 3 in register 2, at least that is what would be expected.